Add serializable shadow roots and getHTML()

annevk opened the pull request:
#### de64e62a2a5041b0d59b985feba7e080993c36b8
<pre>
Add serializable shadow roots and getHTML()
<a href="https://bugs.webkit.org/show_bug.cgi?id=271343">https://bugs.webkit.org/show_bug.cgi?id=271343</a>
<a href="https://rdar.apple.com/125513986">rdar://125513986</a>

Reviewed by Ryosuke Niwa.

This implements <a href="https://github.com/whatwg/html/pull/10139">https://github.com/whatwg/html/pull/10139</a> as amended by
<a href="https://github.com/whatwg/html/pull/10260.">https://github.com/whatwg/html/pull/10260.</a>

This also changes the way a shadow root is serialized to avoid creating
a template element.

The API is guarded by DeclarativeShadowRootsSerializerAPIsEnabled just
in case there is a need to disable it without removing the code.

Thanks to Chris Dumez for his help with Vector&lt;RefPtr&lt;ShadowRoot&gt;&gt;.

WPT tests are synchronized up to and including this commit:
<a href="https://github.com/web-platform-tests/wpt/commit/dd2e51de9985a1c9d381607c2b314b6562545d37">https://github.com/web-platform-tests/wpt/commit/dd2e51de9985a1c9d381607c2b314b6562545d37</a>

This also adjusts the copyright lines to match the commit log for the
relevant files.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-repeats-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-repeats.html:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-ordering-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-ordering.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/w3c-import.log:
* LayoutTests/tests-options.json:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getHTML const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/GetHTMLOptions.h: Added.
* Source/WebCore/dom/GetHTMLOptions.idl: Added.
* Source/WebCore/dom/InnerHTML.idl:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::getHTML const):
(WebCore::ShadowRoot::cloneNodeInternal):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::shouldIncludeShadowRoots const):
(WebCore::MarkupAccumulator::includeShadowRoot const):
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):
(WebCore::MarkupAccumulator::replacementElement):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded):
* Source/WebCore/html/HTMLTemplateElement.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):

Canonical link: <a href="https://commits.webkit.org/277374@main">https://commits.webkit.org/277374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e39889f6b98852ba7ae250e6a133820113acfa36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47429 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50112 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43477 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24069 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38612 "Found 1 new test failure: compositing/repaint/become-overlay-composited-layer.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40877 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19935 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42047 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5472 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/40708 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43768 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51989 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46919 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18792 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45911 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23735 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44948 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24525 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54417 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6682 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23453 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11166 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/editing/markup.h:
```diff
@@ -32,6 +32,7 @@
 #include "HTMLInterchange.h"
 #include "MarkupExclusionRule.h"
 #include "ParserContentPolicy.h"
+#include "ShadowRoot.h"
```
It seems like this should just be `class ShadowRoot`, I agree, but
```
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:1:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/WebCorePrefix.h:165:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashMap.h:25:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTable.h:33:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTraits.h:26:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashFunctions.h:26:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/RefPtr.h:28:
/Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Ref.h:62:16: error: member access into incomplete type 'WebCore::ShadowRoot'
   62 |             ptr->deref();
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:1:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/WebCorePrefix.h:165:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashMap.h:25:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTable.h:33:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashTraits.h:26:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/HashFunctions.h:26:
/Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/RefPtr.h:60:47: note: in instantiation of member function 'WTF::DefaultRefDerefTraits<WebCore::ShadowRoot>::derefIfNotNull' requested here
   60 |     ALWAYS_INLINE ~RefPtr() { RefDerefTraits::derefIfNotNull(PtrTraits::exchange(m_ptr, nullptr)); }
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:1:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/WebCorePrefix.h:168:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/text/AtomString.h:24:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/text/AtomStringImpl.h:23:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/text/UniquedStringImpl.h:28:
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/text/StringImpl.h:35:
/Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Vector.h:70:19: note: in instantiation of member function 'WTF::RefPtr<WebCore::ShadowRoot>::~RefPtr' requested here
   70 |             cur->~T();
/Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Vector.h:253:73: note: in instantiation of member function 'WTF::VectorDestructor<true, WTF::RefPtr<WebCore::ShadowRoot>>::destruct' requested here
  253 |         VectorDestructor<!std::is_trivially_destructible<T>::value, T>::destruct(begin, end);
/Users/annevk/Dev/OpenSource/WebKitBuild/Release/usr/local/include/wtf/Vector.h:781:29: note: in instantiation of member function 'WTF::VectorTypeOperations<WTF::RefPtr<WebCore::ShadowRoot>>::destruct' requested here
  781 |             TypeOperations::destruct(begin(), end());
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:5:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XSLTProcessor.cpp:41:
/Users/annevk/Dev/OpenSource/Source/WebCore/editing/markup.h:106:426: note: in instantiation of member function 'WTF::Vector<WTF::RefPtr<WebCore::ShadowRoot>>::~Vector' requested here
  106 | WEBCORE_EXPORT String serializeFragment(const Node&, SerializedNodes, Vector<Ref<Node>>* = nullptr, ResolveURLs = ResolveURLs::No, std::optional<SerializationSyntax> = std::nullopt, HashMap<String, String>&& replacementURLStrings = { }, HashMap<RefPtr<CSSStyleSheet>, String>&& replacementURLStringsForCSSStyleSheet = { }, IncludeShadowRoots = IncludeShadowRoots::ExplicitOnly, Vector<RefPtr<ShadowRoot>> explicitShadowRoots = { }, const Vector<MarkupExclusionRule>& exclusionRules = { });
In file included from /Users/annevk/Dev/OpenSource/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource443.cpp:1:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XPathVariableReference.cpp:28:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XPathVariableReference.h:28:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XPathExpressionNode.h:29:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XPathValue.h:29:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/xml/XPathNodeSet.h:29:
In file included from /Users/annevk/Dev/OpenSource/Source/WebCore/dom/Node.h:33:
/Users/annevk/Dev/OpenSource/Source/WebCore/dom/TreeScope.h:59:7: note: forward declaration of 'WebCore::ShadowRoot'
   59 | class ShadowRoot;
1 error generated.
```
and `CSSStyleSheet` follows the same pattern.

rniwa replied:
Yeah, ~Vector<RefPtr<ShadowRoot>> will need to see the definition of ShadowRoot.

annevk commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -286,9 +305,19 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
+                        current = firstChild;
+                        namespaceStack.append(namespaceStack.last());
+                        break;
+                    }
```
This is copied from code earlier in this function. Should I put it in some function that returns `Node*` which if non-null we set `current` to or is there a better way of abstracting/sharing this logic?

rniwa commented on Source/WebCore/html/HTMLTemplateElement.cpp:
```diff
@@ -155,10 +155,10 @@ void HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded(Element& host)
+    auto serializable = hasAttributeWithoutSynchronization(HTMLNames::shadowrootserializableAttr);
 
-    auto exceptionOrShadowRoot = host.attachDeclarativeShadow(*mode, delegatesFocus, clonable);
+    auto exceptionOrShadowRoot = host.attachDeclarativeShadow(*mode, delegatesFocus, clonable, serializable);
```
Can we use enum here instead of two booleans? As is, there is no type check done on either value so if someone swaps the order of booleans, the compiler wouldn't catch it.

rniwa commented on Source/WebCore/editing/markup.h:
```diff
@@ -99,10 +100,10 @@ WEBCORE_EXPORT String serializePreservingVisualAppearance(const SimpleRange&, Ve
 String serializePreservingVisualAppearance(const VisibleSelection&, ResolveURLs = ResolveURLs::No, SerializeComposedTree = SerializeComposedTree::No,
     IgnoreUserSelectNone = IgnoreUserSelectNone::Yes, PreserveBaseElement = PreserveBaseElement::No, Vector<Ref<Node>>* = nullptr);
 
+enum class IncludeShadowRoots : uint8_t { ExplicitOnly, Serializable, DeprecatedAll };
```
I don't think `DeprecatedAll` is appropriate naming here. Nothing about this behavior is deprecated.
A crucial distinction is whether we want to serialize all shadow DOM for copy & paste purposes so I'd rather call it `Always` or `All`, or maybe `AllForCopy`.

rniwa commented on Source/WebCore/html/HTMLTemplateElement.idl:
```diff
@@ -37,4 +37,5 @@
     [CEReactions=NotNeeded] attribute [AtomString] DOMString shadowRootMode;
     [CEReactions=NotNeeded, Reflect=shadowrootdelegatesfocus] attribute boolean shadowRootDelegatesFocus;
     [CEReactions=NotNeeded, Reflect=shadowrootclonable] attribute boolean shadowRootClonable;
+    [CEReactions=NotNeeded, Reflect=shadowrootserializable] attribute boolean shadowRootSerializable;    
```
There is a trailing whitespace here.

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -243,6 +267,14 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
     } else
         namespaceStack.constructAndAppend();
 
+    auto getFirstChild = [&](auto& current) -> Node* {
```
"get" prefix should only be used on a function with an out argument.
https://webkit.org/code-style-guidelines/#names-out-argument

rniwa replied:
Maybe `computeFirstChild` or simply `firstChild`?

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -337,21 +373,24 @@ std::pair<String, MarkupAccumulator::IsCreatedByURLReplacement> MarkupAccumulato
+    if (!shadowRoot || !includeShadowRoot(*shadowRoot))
         return nullptr;
 
     auto element = HTMLTemplateElement::create(HTMLNames::templateTag, node.protectedDocument());
```
This is an awfully inefficient way of serializing shadow root.
We should probably cleanup in a follow up to avoid mallocing a phantom element like this.

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -337,21 +373,24 @@ std::pair<String, MarkupAccumulator::IsCreatedByURLReplacement> MarkupAccumulato
     else if (shadowRoot->mode() == ShadowRootMode::Closed)
         element->setShadowRootMode(AtomString { "closed"_s });
+    if (shadowRoot->delegatesFocus())
+        element->setAttribute(HTMLNames::shadowrootdelegatesfocusAttr, emptyAtom());
```
Use `setAttributeWithoutSynchronization` instead.

rniwa commented on Source/WebCore/html/HTMLTemplateElement.cpp:
```diff
@@ -155,10 +156,10 @@ void HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded(Element& host)
+    auto serializable = hasAttributeWithoutSynchronization(HTMLNames::shadowrootserializableAttr);
 
-    auto exceptionOrShadowRoot = host.attachDeclarativeShadow(*mode, delegatesFocus, clonable);
+    auto exceptionOrShadowRoot = host.attachDeclarativeShadow({ *mode, delegatesFocus, clonable, serializable });
```
I don't think this will address my concern that the orders of delegatesFocus, clonable, & serializable can change without any type failures.
We should define 3 enum classes like these:
```
enum class ShadowRootDelegatesFocus : bool { No, Yes };
enum class ShadowRootIsClonable : bool { No, Yes };
enum class ShadowRootIsSerializable : bool { No, Yes }
```

annevk commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -56,6 +56,7 @@
 #include "SVGElementInlines.h"
 #include "Settings.h"
 #include "ShadowRoot.h"
+#include "ShadowRootInit.h"
```
This needs to be removed.

annevk commented on Source/WebCore/html/HTMLTemplateElement.cpp:
```diff
@@ -143,22 +144,18 @@ void HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded(Element& host)
+    auto mode = modeString == "closed"_s ? ShadowRootMode::Closed : ShadowRootMode::Open;
 
-    auto clonable = hasAttributeWithoutSynchronization(HTMLNames::shadowrootclonableAttr);
+    auto delegatesFocus = hasAttributeWithoutSynchronization(HTMLNames::shadowrootdelegatesfocusAttr)  ? ShadowRootDelegatesFocus::Yes : ShadowRootDelegatesFocus::No;
```
Space before `?` needs to be removed.

annevk reviewed (commented):
Found some nits, but not pushing again as seems better to wait for review.

rniwa commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright (C) 2011 Google Inc. All rights reserved.
+∂ * Copyright (C) 2013-2024 Apple Inc. All rights reserved.
```
Revert?

rniwa commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -335,34 +371,43 @@ std::pair<String, MarkupAccumulator::IsCreatedByURLReplacement> MarkupAccumulato
+        case ShadowRootMode::Closed:
+            m_markup.append("closed"_s);
+            break;
+        default:
```
We should explicitly list ShadowRootMode::UserAgent instead.

annevk committed:
Add serializable shadow roots and getHTML()
https://bugs.webkit.org/show_bug.cgi?id=271343
rdar://125513986

Reviewed by Ryosuke Niwa.

This implements https://github.com/whatwg/html/pull/10139 as amended by
https://github.com/whatwg/html/pull/10260.

This also changes the way a shadow root is serialized to avoid creating
a template element.

The API is guarded by DeclarativeShadowRootsSerializerAPIsEnabled just
in case there is a need to disable it without removing the code.

Thanks to Chris Dumez for his help with Vector<RefPtr<ShadowRoot>>.

WPT tests are synchronized up to and including this commit:
https://github.com/web-platform-tests/wpt/commit/dd2e51de9985a1c9d381607c2b314b6562545d37

This also adjusts the copyright lines to match the commit log for the
relevant files.

* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-repeats-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-repeats.html:
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-ordering-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml-ordering.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/gethtml.tentative.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/w3c-import.log:
* LayoutTests/tests-options.json:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getHTML const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/GetHTMLOptions.h: Added.
* Source/WebCore/dom/GetHTMLOptions.idl: Added.
* Source/WebCore/dom/InnerHTML.idl:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::getHTML const):
(WebCore::ShadowRoot::cloneNodeInternal):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/ShadowRoot.idl:
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::MarkupAccumulator):
(WebCore::MarkupAccumulator::shouldIncludeShadowRoots const):
(WebCore::MarkupAccumulator::includeShadowRoot const):
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):
(WebCore::MarkupAccumulator::replacementElement):
* Source/WebCore/editing/MarkupAccumulator.h:
(WebCore::MarkupAccumulator::MarkupAccumulator):
* Source/WebCore/editing/markup.cpp:
(WebCore::serializeFragment):
* Source/WebCore/editing/markup.h:
(WebCore::serializeFragment):
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded):
* Source/WebCore/html/HTMLTemplateElement.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):

Canonical link: https://commits.webkit.org/277374@main
